File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/utils/PropertyPlaceholder.java
package org.xyattic.eventual.consistency.support.core.utils;

import lombok.extern.slf4j.Slf4j;
import org.springframework.core.env.MutablePropertySources;
import org.springframework.core.env.PropertyResolver;
import org.springframework.core.env.PropertySourcesPropertyResolver;
import org.springframework.lang.Nullable;

@Slf4j
public class PropertyPlaceholder {

    @Nullable
    public static String getProperty(final String key) {
        return getPropertyResolver().getProperty(key);
    }

    public static boolean containsProperty(final String key) {
        return getPropertyResolver().containsProperty(key);
    }

    public static String getProperty(final String key, final String defaultValue) {
        return getPropertyResolver().getProperty(key, defaultValue);
    }

    @Nullable
    public static <T> T getProperty(final String key, final Class<T> targetType) {
        return getPropertyResolver().getProperty(key, targetType);
    }

    public static <T> T getProperty(final String key, final Class<T> targetType, final T defaultValue) {
        return getPropertyResolver().getProperty(key, targetType, defaultValue);
    }

    public static String getRequiredProperty(final String key) throws IllegalStateException {
        return getPropertyResolver().getRequiredProperty(key);
    }

    public static <T> T getRequiredProperty(final String key, final Class<T> targetType) throws IllegalStateException {
        return getPropertyResolver().getRequiredProperty(key, targetType);
    }

    public static String resolvePlaceholders(final String text) {
        return getPropertyResolver().resolvePlaceholders(text);
    }

    public static String resolveRequiredPlaceholders(final String text) throws IllegalArgumentException {
        return getPropertyResolver().resolveRequiredPlaceholders(text);
    }

    private static PropertyResolver getPropertyResolver() {
        try {
            return SpringContextUtils.getEnvironment();
        } catch (Exception e) {
            log.warn("无法获取spring环境变量");
            return new PropertySourcesPropertyResolver(new MutablePropertySources());
        }
    }

}

File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/utils/PropertyPlaceholder.kt
package org.xyattic.eventual.consistency.support.core.utils

import org.springframework.core.env.MutablePropertySources
import org.springframework.core.env.PropertyResolver
import org.springframework.core.env.PropertySourcesPropertyResolver
import org.springframework.lang.Nullable

object PropertyPlaceholder {

    private val log = getLogger()

    @Nullable
    fun getProperty(key: String): String? {
        return propertyResolver.getProperty(key)
    }

    fun containsProperty(key: String): Boolean {
        return propertyResolver.containsProperty(key)
    }

    fun getProperty(key: String, defaultValue: String): String {
        return propertyResolver.getProperty(key, defaultValue)
    }

    @Nullable
    fun <T> getProperty(key: String, targetType: Class<T>): T? {
        return propertyResolver.getProperty(key, targetType)
    }

    fun <T> getProperty(key: String, targetType: Class<T>, defaultValue: T): T {
        return propertyResolver.getProperty(key, targetType, defaultValue)
    }

    fun getRequiredProperty(key: String): String {
        return propertyResolver.getRequiredProperty(key)
    }

    fun <T> getRequiredProperty(key: String, targetType: Class<T>): T {
        return propertyResolver.getRequiredProperty(key, targetType)
    }

    fun resolvePlaceholders(text: String): String {
        return propertyResolver.resolvePlaceholders(text)
    }

    fun resolveRequiredPlaceholders(text: String): String {
        return propertyResolver.resolveRequiredPlaceholders(text)
    }

    private val propertyResolver: PropertyResolver
        get() = try {
            SpringContextUtils.environment
        } catch (e: Exception) {
            log.warn("无法获取spring环境变量")
            PropertySourcesPropertyResolver(MutablePropertySources())
        }
}